cpp/crypto/CipherContext.cpp

```diff
@@ -139,81 +139,151 @@ CipherContext &
+        throw;
+      }
+
+      // default iv len for AES is 96 bits but driver's IV generator
+      // uses 128 bits defined as SF_CRYPTO_IV_NBITS
+      if (1 != EVP_CIPHER_CTX_ctrl(m_pimpl->ctx, EVP_CTRL_GCM_SET_IVLEN, SF_CRYPTO_IV_NBITS / 8, nullptr)) {
+        CXX_LOG_ERROR("Failed to set IV len equal to %d bits", SF_CRYPTO_IV_NBITS);
+        throw;
+      }
+
+      // initialize key and iv
+      if (1 != EVP_EncryptInit_ex(m_pimpl->ctx, nullptr, nullptr,
+        reinterpret_cast<const unsigned char *>(m_pimpl->key->data),
+        reinterpret_cast<const unsigned char *>(m_pimpl->iv.data))) {
+        CXX_LOG_ERROR("Failed to initialize encryption key and iv")
+        throw;
+      }
+
+    break;
+    default:
+      if (1 != EVP_EncryptInit_ex(m_pimpl->ctx, m_pimpl->cipher, nullptr,
+         reinterpret_cast<const unsigned char *>(m_pimpl->key->data),
+         reinterpret_cast<const unsigned char *>(m_pimpl->iv.data))) {
+        CXX_LOG_ERROR("Failed to initialize encryption encryption operation with key and iv")
+        throw;
+      }
+    break;
+  }
+}
+
+void CipherContext::initialize_decryption() const {
+  CXX_LOG_DEBUG("Initializing decryption, mode = %d", m_pimpl->mode);
+  switch (m_pimpl->mode) {
+    case CryptoMode::GCM:
+      if (1 != EVP_DecryptInit_ex(m_pimpl->ctx, m_pimpl->cipher, nullptr, nullptr, nullptr)) {
+        CXX_LOG_ERROR("Failed to initialize decryption operation for provided cipher");
+        throw;
+      }
+
+      if (1 != EVP_CIPHER_CTX_ctrl(m_pimpl->ctx, EVP_CTRL_GCM_SET_IVLEN, SF_CRYPTO_IV_NBITS / 8, nullptr)) {
+        CXX_LOG_ERROR("Failed to set IV len equal to %d bits", SF_CRYPTO_IV_NBITS);
+        throw;
+      }
+
+      if (1 != EVP_DecryptInit_ex(m_pimpl->ctx, nullptr, nullptr,
+        reinterpret_cast<const unsigned char *>(m_pimpl->key->data),
+        reinterpret_cast<const unsigned char *>(m_pimpl->iv.data))) {
+        CXX_LOG_ERROR("Failed to initialize decryption key and iv")
+        throw;
+      }
+    break;
+    default:
+      if (1 != EVP_DecryptInit_ex(m_pimpl->ctx, m_pimpl->cipher, nullptr,
+         reinterpret_cast<const unsigned char *>(m_pimpl->key->data),
+         reinterpret_cast<const unsigned char *>(m_pimpl->iv.data)))
+    break;
+  }
+}
+
+
+
+void CipherContext::set_padding() const {
+  // EVP_CIPHER_CTX_set_padding() must be called after initialize
+  // https://www.openssl.org/docs/manmaster/man3/EVP_CIPHER_CTX_set_padding.html
+  switch (m_pimpl->padding) {
+    case CryptoPadding::NONE:
+      break;
+    case CryptoPadding::PKCS5:
+      if (1 != EVP_CIPHER_CTX_set_padding(m_pimpl->ctx, true)) {
+        CXX_LOG_ERROR("Failed to set padding to PKCS5")
+        throw;
+      }
+    break;
+  }
+}
+
+void CipherContext::set_aad(const unsigned char *aad, int aad_len) const {
+  if (m_pimpl->mode != CryptoMode::GCM) {
+    CXX_LOG_ERROR("AAD is not supported outside GCM mode");
+    throw;
+  }
+  int len;
```

Comment:
Actually I believe it may be a `nullptr`. What this variable is for is capturing the amount of bytes written so it should be equal to `aad_len`. I was thinking about returning this value but since we're throwing anyway - it's no longer required